File: 0148-sort-list/0148-sort-list.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode sortList(ListNode head) {
        if(head==null || head.next==null){
            return head;
        }
        ListNode mid=mid(head);
        ListNode left=sortList(head);
        ListNode right=sortList(mid);
        return merge(left,right);
    }
    private ListNode mid(ListNode head){
        ListNode i=head ,j=head ,prev=null;
        while(j!=null &&j.next!=null){
            prev=i;
            i=i.next;
            j=j.next.next;
        }
        if(prev!=null)prev.next=null;
        return i;
    }
    private ListNode merge(ListNode l1,ListNode l2){
        ListNode dummy=new ListNode(0);
        ListNode t=dummy;
        while(l1!=null && l2!=null){
            if(l1.val<l2.val){
                t.next=l1;
                l1=l1.next;
            }
            else{
                t.next=l2;
                l2=l2.next;
            }
            t=t.next;
        }
        t.next=(l1!=null)?l1:l2;
        return dummy.next;
    }

}

File: 0148-sort-list/0148-sort-list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def sortList(self, head: Optional[ListNode]) -> Optional[ListNode]:
        if not head or not head.next:
            return head
        slow=head
        fast=head.next
        while fast and fast.next:
            slow=slow.next
            fast=fast.next.next
        mid=slow.next
        slow.next=None
        left=self.sortList(head)
        right=self.sortList(mid)
        return self.merge(left,right)
    def merge(self,left,right):
        dummy=ListNode(0)
        temp=dummy
        while left and right:
            if left.val<right.val:
                temp.next=left
                left=left.next
            else:
                temp.next=right
                right=right.next
            temp=temp.next
        while left:
            temp.next=left
            left=left.next
            temp=temp.next
        while right:
            temp.next=right
            right=right.next
            temp=temp.next
        return dummy.next